File: src/main/java/com/sat/common/exhandler/CommonRestControllerAdvice.java
package com.sat.common.exhandler;

import com.sat.common.dto.ErrorResponse;
import com.sat.common.exception.DataNotFoundException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.BindException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.time.format.DateTimeParseException;

@Slf4j
@RestControllerAdvice
public class CommonRestControllerAdvice {

    @ExceptionHandler(BindException.class)
    public ResponseEntity<ErrorResponse> error(BindException e) {
        ErrorResponse errorResponse = ErrorResponse.of(e);
        return ResponseEntity.badRequest().body(errorResponse);
    }

    @ExceptionHandler(DateTimeParseException.class)
    public ResponseEntity<ErrorResponse> error(DateTimeParseException e) {
        ErrorResponse errorResponse = ErrorResponse.of(e);
        return ResponseEntity.badRequest().body(errorResponse);
    }

    @ExceptionHandler(DataNotFoundException.class)
    public ResponseEntity<ErrorResponse> error(DataNotFoundException e) {
        ErrorResponse errorResponse = ErrorResponse.of(e);
        return ResponseEntity.badRequest().body(errorResponse);
    }

//    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<ErrorResponse> error(IllegalArgumentException e) {
        ErrorResponse errorResponse = ErrorResponse.of(e);
        return ResponseEntity.badRequest().body(errorResponse);
    }
}

File: src/main/kotlin/com/sat/common/ui/web/CommonRestControllerAdvice.kt
package com.sat.common.ui.web

import com.sat.common.domain.exception.NotFoundException
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.beans.propertyeditors.StringTrimmerEditor
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.WebDataBinder
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.InitBinder
import org.springframework.web.bind.annotation.RestControllerAdvice

val log = KotlinLogging.logger { }

@RestControllerAdvice
class CommonRestControllerAdvice {

    @InitBinder
    fun initBinder(binder: WebDataBinder) {
        val editor = StringTrimmerEditor(true)
        binder.registerCustomEditor(String::class.java, editor)
    }

    @ExceptionHandler(IllegalArgumentException::class, IllegalStateException::class)
    fun error(e: RuntimeException): ResponseEntity<ErrorResponse> {
        return ResponseEntity.badRequest().body(ErrorResponse(e.message ?: ""))
    }

    @ExceptionHandler(NotFoundException::class)
    fun error(e: NotFoundException): ResponseEntity<ErrorResponse> {
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(ErrorResponse(e.message))
    }

    @ExceptionHandler(Exception::class)
    fun error(e: Exception): ResponseEntity<ErrorResponse> {
        log.error { e.message }
        return ResponseEntity.internalServerError().body(ErrorResponse("에러 발생"))
    }
}
